### PR TITLE
Support wazuh in the ci-multinode environment

### PR DESCRIPTION
**Pull request by Alex-Welsh**

### CODE REVIEW

**Comment by Alex-Welsh** [comment]
Matching terraform config PR [here](https://github.com/stackhpc/terraform-kayobe-multinode/pull/26)

File: doc/source/configuration/wazuh.rst [hunk 1]
```diff
@@ -74,7 +89,7 @@ Define network interfaces ``etc/kayobe/inventory/group_vars/wazuh-manager/networ
 
 
 The Wazuh manager may need to be exposed externally, in which case it may require another interface.
-This can be done as follows in ``etc/kayobe/inventory/group_vars/wazuh-manager/network-interfaces`` ,
+This can be done as follows in ``etc/kayobe/inventory/group_vars/wazuh-manager/network-interfaces``,
 with the network defined in network.yml as usual.
```
Comment (MoteHue) [inline doc/source/configuration/wazuh.rst]:
```suggestion
with the network defined in ``networks.yml`` as usual.
```

File: doc/source/configuration/wazuh.rst [hunk 2]
```diff
@@ -210,7 +225,7 @@ You may need to modify some of the variables, including:
 .. note::
 
     NOTE:
-    If you are using multiple environments, and you need to customise Wazuh in each environement, create override files in an appropriate directory,
+    If you are using multiple environments, and you need to customise Wazuh in each environment, create override files in an appropriate directory,
     for example `etc/kayobe/environments/production/inventory/group_vars/`
     Files which values can be overridden (in context of Wazuh):
```
Comment (MoteHue) [inline doc/source/configuration/wazuh.rst]:
```suggestion
    Files which values can be overridden (in the context of Wazuh):
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 3]
```diff
@@ -324,3 +324,66 @@ with:
 
 There are various other options for sonobuoy, see the `documentation
 <https://sonobuoy.io/docs/>`_ for more details.
+
+Wazuh
+======
+
+Wazuh is supported but not deployed by default. If you are using the
+standard StackHPC multinode terraform, there is a ``deploy_wazuh`` terraform
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
standard [StackHPC multinode terraform](https://github.com/stackhpc/terraform-kayobe-multinode), there is a ``deploy_wazuh`` terraform
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 4]
```diff
@@ -324,3 +324,66 @@ with:
...
+Add the hostname to the ``[wazuh-manager]`` group in
+``$KAYOBE_CONFIG_PATH/environments/ci-multinode/inventory/hosts``.
+
+Create the following directory structure:
+``$KAYOBE_CONFIG_PATH/hooks/infra-vm-host-configure/pre.d/``
+
+Either copy or symlink in the growroot, networking, and vxlan playbooks as
+shown in ``$KAYOBE_CONFIG_PATH/hooks/seed-host-configure/pre.d/``.
+
+Configure the wazuh manager vm:
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
Configure the Wazuh manager VM:
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 5]
```diff
@@ -324,3 +324,66 @@ with:
...
+
+.. code-block:: bash
+
+      kayobe infra vm host configure
+
+Create and encrypt the wazuh secrets
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
Create and encrypt the Wazuh secrets
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 6]
```diff
@@ -324,3 +324,66 @@ with:
...
+
+.. code-block:: bash
+
+      kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-secrets.yml
+      ansible-vault encrypt --vault-password-file ~/vault.password  $KAYOBE_CONFIG_PATH/environments/ci-multinode/wazuh-secrets.yml
+
+Run the wazuh manager and agent deployment playbooks:
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
Run the Wazuh manager and agent deployment playbooks:
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 7]
```diff
@@ -324,3 +324,66 @@ with:
...
+
+.. code-block:: bash
+
+      kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-manager.yml
+      kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-agent.yml
+
+Wazuh should now be fully deployed. To test the service, you can use sshuttle
+or some other forwarding protocol to access the wazuh dashboard e.g.
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
or some other forwarding protocol to access the Wazuh dashboard e.g.
```

File: doc/source/contributor/environments/ci-multinode.rst [hunk 8]
```diff
@@ -324,3 +324,66 @@ with:
...
+
+The default username is ``admin`` and the password is the
+``opendistro_admin_password`` which can be found in ``wazuh-secrets.yml`` e.g.
+
+.. code-block:: bash
+
+      ansible-vault view $KAYOBE_CONFIG_PATH/environments/ci-multinode/wazuh-secrets.yml --vault-password-file ~/vault.password | grep opendistro_admin_password
+
+If the deployment has been successful, you should be able to see a wazuh agent
+for each host in your deployment (aside from the wazuh manager itself).
```
Comment (MoteHue) [inline doc/source/contributor/environments/ci-multinode.rst]:
```suggestion
If the deployment has been successful, you should be able to see a Wazuh agent
for each host in your deployment (aside from the Wazuh manager itself).
```

File: releasenotes/notes/multinode-wazuh-c9edd65ab5cdbcb8.yaml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Added support for wazuh in the ci-multinode environment.
```
Comment (MoteHue) [inline releasenotes/notes/multinode-wazuh-c9edd65ab5cdbcb8.yaml]:
```suggestion
    Added support for Wazuh in the ci-multinode environment.
```

**Review by MoteHue — CHANGES_REQUESTED** [review]

File: doc/source/configuration/wazuh.rst (hunk 1 above)
Comment (Alex-Welsh) [reply doc/source/configuration/wazuh.rst]: Done

File: doc/source/configuration/wazuh.rst (hunk 2 above)
Comment (Alex-Welsh) [reply doc/source/configuration/wazuh.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 3 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 4 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 5 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 6 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 7 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: doc/source/contributor/environments/ci-multinode.rst (hunk 8 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: Done

File: releasenotes/notes/multinode-wazuh-c9edd65ab5cdbcb8.yaml (hunk 9 above)
Comment (Alex-Welsh) [reply releasenotes/notes/multinode-wazuh-c9edd65ab5cdbcb8.yaml]: Done

**Review by MoteHue — DISMISSED** [review]

File: doc/source/configuration/wazuh.rst [hunk 10]
```diff
@@ -2,6 +2,21 @@
 Wazuh
 =====
 
+The short version
```
Comment (markgoddard) [inline doc/source/configuration/wazuh.rst:5]: Is this intended as an introduction, or to be complete, for people familiar with the procedure?

File: doc/source/contributor/environments/ci-multinode.rst [hunk 11]
```diff
@@ -324,3 +324,67 @@ with:
...
+
+Wazuh
+======
+
+Wazuh is supported but not deployed by default. If you are using the standard
+[StackHPC multinode
+terraform](https://github.com/stackhpc/terraform-kayobe-multinode), there is a
+``deploy_wazuh`` terraform variable that will add it to the automated setup.
+
+This guide will explain how to add Wazuh to an existing multinode environment.
```
Comment (markgoddard) [inline doc/source/contributor/environments/ci-multinode.rst]: Make it clear that this is done automatically by the script when using terraform?

File: etc/kayobe/ansible/configure-vxlan.yml [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ---
 - name: Configure VXLAN
-  hosts: controllers,compute,seed,storage
+  hosts: "{{ configure_vxlan_playbook_hosts | default('seed,compute,controllers,storage') }}"
```
Comment (markgoddard) [inline etc/kayobe/ansible/configure-vxlan.yml]: Why not add infra-vms to the default?

File: etc/kayobe/ansible/fix-networking.yml [hunk 13]
```diff
@@ -1,6 +1,6 @@
 ---
 - name: Fix networking
-  hosts: seed,compute,controllers,storage
+  hosts: "{{ fix_networking_playbook_hosts | default('seed,compute,controllers,storage') }}"
```
Comment (markgoddard) [inline etc/kayobe/ansible/fix-networking.yml]: ditto

File: doc/source/configuration/wazuh.rst (hunk 10 above)
Comment (Alex-Welsh) [reply doc/source/configuration/wazuh.rst:5]: Intended as a short version for people who are familar with it but perhaps need a reminder.

File: doc/source/contributor/environments/ci-multinode.rst (hunk 11 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: My intention was that the [previous paragraph](https://github.com/stackhpc/stackhpc-kayobe-config/pull/552/files#diff-6c737a95f8f4930660d780c0727646e75c4834280d2515b617d87855e0ea61e8R331-R334) would explain that it can be done automatically with terraform, but this was an alternative. I can reword it to make it more clear if you think that would be best.

File: etc/kayobe/ansible/configure-vxlan.yml (hunk 12 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/configure-vxlan.yml]: I wanted to keep the playbook as general as possible. I assumed that in a normal deployment we might want to treat infrastructure VMs differently and only configure the other hosts.

Comment (markgoddard) [reply etc/kayobe/ansible/configure-vxlan.yml]: This playbook is just for multinode, so don't need to be too careful.

File: doc/source/contributor/environments/ci-multinode.rst (hunk 11 above)
Comment (markgoddard) [reply doc/source/contributor/environments/ci-multinode.rst]: It's clear if you read it carefully... How about a subheading to separate it from the first paragraph?

Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst]: I've separated the sections out so now we have two subheadings, one for new deployments, one for existing

File: etc/kayobe/ansible/configure-vxlan.yml (hunk 12 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/configure-vxlan.yml]: I've reverted the change and just added the group

File: etc/kayobe/ansible/fix-networking.yml (hunk 13 above)
Comment (Alex-Welsh) [reply etc/kayobe/ansible/fix-networking.yml]: I've reverted the change and just added the group

File: doc/source/contributor/environments/ci-multinode.rst [hunk 14]
```diff
@@ -324,3 +324,71 @@ with:
...
+``$KAYOBE_CONFIG_PATH/hooks/infra-vm-host-configure/pre.d/``
+
+Either copy or symlink in the growroot, networking, and vxlan playbooks as
+shown in ``$KAYOBE_CONFIG_PATH/hooks/seed-host-configure/pre.d/``.
+
+Configure the Wazuh manager VM:
+
+.. code-block:: bash
+
+      kayobe infra vm host configure
```
Comment (markgoddard) [inline doc/source/contributor/environments/ci-multinode.rst:363]: For this to work the host also needs to be in the infra-vms group.

File: doc/source/contributor/environments/ci-multinode.rst [hunk 15]
```diff
@@ -324,3 +324,71 @@ with:
...
+
+      kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-manager.yml
+      kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/wazuh-agent.yml
+
+Wazuh should now be fully deployed. To test the service, you can use sshuttle
+or some other forwarding protocol to access the Wazuh dashboard e.g.
+
+.. code-block:: bash
+
+      sshuttle -r <wazuh-manager-hostname> <wazuh-manager-ip>
```
Comment (markgoddard) [inline doc/source/contributor/environments/ci-multinode.rst:384]: Are these arguments correct? I'm not really familiar with sshuttle

Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst:384]: They do work, but you do need to have some SSH config set up to the point where you can run ``ssh <wazuh-manager-hostname>`` and it'll work. I'll add a bit more explanation to it

File: doc/source/contributor/environments/ci-multinode.rst (hunk 14 above)
Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst:363]: That's true. The group heirachy is created in terraform regardless of whether deploy_wazuh is true or false, but it won't apply to multinodes (or branches) that exist from before the changes.

Comment (Alex-Welsh) [reply doc/source/contributor/environments/ci-multinode.rst:363]: Fixed

**Review by markgoddard — APPROVED** [review]
Good stuff
